Support environment variables in odoo.conf

codebykyle opened the pull request:
Following on from the changes mentioned in:
https://github.com/odoo/docker/issues/538

This also resolves:
https://github.com/odoo/docker/issues/535

and permanently addresses:
https://github.com/odoo/docker/pull/497

codebykyle commented:
To test the changes, cd into the `18.0` directory. You can use the included `docker-compose.yml` to build the docker container and launch a copy of Odoo using these changes.

```sh
cd 18.0
docker compose build && docker compose up
```

codebykyle commented:
I'm further testing this. 

One thing that I do not like is that if you mount a config file at /etc/odoo/odoo.conf, it will be overwritten once merged on the host machine. This is because the mounted file is the same location as the config file that Odoo uses.

Additionally, the file will be mounted as `ubuntu:ubuntu`. This means that we need to add Odoo to the `ubuntu` group via `RUN usermod -aG ubuntu odoo`.

The ps python script can remain as-is.

There are two ways to address this:

## Option 1: Change ODOO_RC to another location

We can use the default /etc/odoo/odoo.conf as the input for envsubst. We update the ODOO_RC env location in the docker file. This will enable these features by default. No existing changes would need to be done for setups mounting the configuration in /etc/odoo/odoo.conf, we would just natively support this

## Option 2: Make this feature optional using a .template.conf file

Moving the file into a different file other than `/etc/odoo/odoo.conf` would make this feature optional. For example, if the user mounts a file at `/etc/odoo-docker/odoo.conf`, we generate the `/etc/odoo/odoo.conf`. If there is no file there, we skip this step and do not enable environment variable injection. This would be a default off, optional method of doing this.

codebykyle commented:
I've pushed a change to address that issue. 

The config file that Odoo now loads is `/etc/odoo/odoo_docker.conf`, as specified in ODOO_RC in the `Dockerfile`. 

The wait script automatically picked that change up. 

This enables this feature by default, and reduces any complexity that would come with template files. 

This also is a drop-in replacement for current builds, nothing will need to change for end users to take advantage of this. If the user mounts their config to /etc/odoo/odoo.conf, under the hood, it will automatically be moved to `/etc/odoo/odoo_docker.conf` as-is.

codebykyle commented:
Further testing:

Forgot to add the new environment variables to match the Odoo config values, eg:

`DB_HOST` as opposed to HOST.

The existing environment variables should continue to work (eg, HOST), but they are mapped to $DB_HOST for consistency. 

For example:

```yml
    environment:
#      DB_HOST: ${DB_HOST:-db}
      HOST: "test"
```

will be merged such that the following:
```.conf
[options]
db_host = $DB_HOST
```

results in a correct merge:

![image](https://github.com/user-attachments/assets/c62795d8-a112-440a-b804-4c6e955b9d97)

However, the `DB_HOST `, `DB_PORT`, etc are the ideal settings as they match the config name

codebykyle commented:
I tested all the config settings. 

The only one that had an issue was `osv_memory_age_limit`, which has been renamed to `transient_age_limit`, per the warning generated by Odoo.

`2025-02-11 06:33:25,600 1 WARNING ? py.warnings: /usr/lib/python3/dist-packages/odoo/tools/config.py:603: DeprecationWarning: The 'osv_memory_age_limit' option found in the configuration file is a deprecated alias to 'transient_age_limit', please use the latter.`

I've pushed a change to rename that option. 

I think most major development efforts with this are done. Some testing should be performed by someone other than me. All my needs are more or less met. I think one minor change I may make is to make the timeout configurable by `env`, but that isn't critical for my use-case.

If we want to continue with this method, I can get some documentation up for the settings. Please let me know if anyone has any feedback on this.

amh-mw commented:
> Sorry, I figured if we're going to do it, may as well go all the way with it.

AFAIK this repository has no maintainer yet receives bi-weekly updates from some internal Odoo SA build process. I have never seen a PR get merged.

codebykyle commented:
That isn't surprising and more or less checks out. This repository as it is does not support many features of Odoo, specifically around multi-tenancy. I'll leave these changes here, as they work, and anyone can use them, or I will create my own container with these features and publish that to Dockerhub, whatever that is worth.

These are necessary features, especially if you are selling or hosting Odoo.

codebykyle commented:
I have submitted a patch to fix a potential race condition with ps_isready on initial startup. That should be resolved.

This will likely be the last PR update I submit here if this repository is not maintained. Having checked the commit history, I believe you are correct @amh-mw 

I actually do need this container to be working this way, so I have moved my efforts over to a new repository.

[It is available here.](https://github.com/adomi-io/odoo-docker)

I'll accept PRs if anyone would like to contribute.

I also switched the image over to Alpine Linux and split this up into a multi-layer build process for things like wkhtmltopdf, node, and source management, so that we're able to actually automate the build process with Github actions, as opposed to some internal build process over at Odoo. The resulting docker file is very lightweight by comparison to the full Ubuntu image this repo is pulling.

This should let us setup a mirror of the Odoo Github repo so we are not dependent on some internal build process to submit secret updates that occasionally break enterprise. I'm happy to contribute those changes back to this repository but it doesnt seem like that is going to be accepted, so why bother. 

In the meantime, the changes submitted here should work if anyone wants these features. Let me know if anyone out there in the void ends up using this.

Thanks guys.
